src/Lumper.Lib/BSP/Struct/PakFileEntry.cs

```diff
@@ -7,50 +7,98 @@ namespace Lumper.Lib.BSP.Struct;
+/// We assume one is non-null throughout.
+/// </summary>
+public sealed class PakfileEntry : IDisposable
 {
-    public PakFileEntry(ZipArchiveEntry entry)
+    public PakfileEntry(ZipArchiveEntry zipEntry)
     {
-        _entry = entry;
-        Key = entry.Key;
+        ZipEntry = zipEntry;
+        Key = zipEntry.Key ?? throw new InvalidDataException("Pakfile contains an item without a key");
     }
 
-    public PakFileEntry(string key, Stream stream)
+    public PakfileEntry(string key, Stream stream)
     {
         Key = key;
-        _dataStream = stream;
+
+        stream.Seek(0, SeekOrigin.Begin);
+        var mem = new MemoryStream();
+        stream.CopyTo(mem);
+        _buffer = mem.ToArray();
     }
 
+    // Probably shouldn't be public but whatever. be careful!
     [JsonIgnore]
-    private readonly ZipArchiveEntry? _entry;
+    public ZipArchiveEntry? ZipEntry { get; }
 
     public string Key { get; set; }
 
+    [JsonProperty]
+    public long? CompressedSize => ZipEntry?.CompressedSize ?? null;
+
     [JsonIgnore]
-    private Stream? _dataStream;
+    public bool IsModified { get; set; }
+
+    [JsonIgnore]
+    private byte[]? _buffer;
+
+    private readonly List<MemoryStream> _issuedStreams = [];
 
     [JsonIgnore]
-    public Stream DataStream
+    private readonly object _lock = new();
+
+    /// <summary>
+    /// Get an stream to the uncompressed pakfile entry.
+    ///
+    /// This stream can be invalidated if the entry is updated by the user.
+    /// </summary>
+    public MemoryStream GetReadOnlyStream()
+    {
+        lock (_lock)
+        {
+            if (_buffer is not null)
+                return IssueStream();
+
+            using Stream zipStream = ZipEntry!.OpenEntryStream();
+            using var outStream = new MemoryStream();
+            zipStream.CopyTo(outStream);
+            _buffer = outStream.ToArray();
+            return IssueStream();
+        }
+    }
+
+    private MemoryStream IssueStream()
     {
-        get => _dataStream ?? _entry!.OpenEntryStream(); // Ctors take either entry or stream, one is non-null
-        set
+        var stream = new MemoryStream(_buffer!, writable: false);
+        _issuedStreams.Add(stream);
+        return stream;
+    }
+
+    public void UpdateData(byte[] data)
+    {
+        lock (_lock)
         {
-            if (_dataStream is not null)
-            {
-                _dataStream.Close();
-                _dataStream.Dispose();
-            }
-            _dataStream = value;
+            DisposeIssuedStreams();
+            _buffer = data;
         }
     }
 
-    public bool IsModified => _dataStream != null;
+    private void DisposeIssuedStreams()
+    {
+        foreach (MemoryStream oldStream in _issuedStreams)
+            oldStream.Close();
+    }
+
+    public void Dispose() => DisposeIssuedStreams();
 
+    // Doesn't matter that this is unused, is used for JSON comparisons.
```

Comment:
could just say its used for the json dump

Reply:
We don't have a particular reason for using MD5 do we? My game asset removal PR needs a SHA1 version of this, so I just replaced the MD5, seemed unnecessary having both.

Reply:
no reason for md5, just picked something cheap